server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -409,13 +443,9 @@ public void applyClusterState(ClusterChangedEvent event) {
                     if (event.routingTableChanged() && waitingShardsStartedOrUnassigned(snapshotsInProgress, event)) {
                         processStartedShards();
                     }
-                    // Cleanup all snapshots that have no more work left:
-                    // 1. Completed snapshots
-                    // 2. Snapshots in state INIT that a previous master of an older version failed to start
-                    // 3. Snapshots in any other state that have all their shard tasks completed
-                    snapshotsInProgress.entries().stream().filter(
```

Comment:
Running this check on every CS update wasn't great and added a lot of cycles on the CS thread for larger snapshots. It would be even worse when we start actually having multiple snapshot-in-progress entries. We really should only have to do a full check if anything changed about the entries' shards or on master fail-over.
This is change moves all the updating of the shard entries into this class so we can selectively run this check (though see my other comment below, we could be even more selective in a follow-up :)).